### PR TITLE
Add type and buffer for delta qp per block

### PR DESCRIPTION
**Pull request by Tianhaol**

### CODE REVIEW

**Comment by Tianhaol** [comment]
@XinfengZhang  Could you help add reviewers and review it?

**Review by XinfengZhang — CHANGES_REQUESTED** [review]

File: va/va_enc_hevc.h [hunk 1]
```diff
@@ -140,8 +140,12 @@ typedef union VAConfigAttribValEncHEVCFeatures {
...
          */
         uint32_t deblocking_filter_disable  : 2;
-        /* Reserved,should not be used, avoid conflict with VA_ATTRIB_NOT_SUPPORTED. */
-        uint32_t reserved                   : 2;
+        /** Delta qp or qp per block enable.
+         *
+         * Allows setting delta qp or qp per block
+         * 1: qp enabled 2: delta qp enabled 3: both enabled.
+         */
+        uint32_t qp_per_block_enable        : 2;
```
[the remark this answers is not in the document]
Comment (Tianhaol) [reply va/va_enc_hevc.h]: I do not use top bit, just use the last bit.

**Comment by Tianhaol** [comment]
@XinfengZhang  Could you please this PR again?

File: va/va.h [hunk 2]
```diff
@@ -2044,6 +2044,15 @@ typedef enum {
...
     VAEncryptionParameterBufferType = 60,
 
+    /**
+     * \brief Encoding Delta QP buffer
+     *
+     * This buffer contains delta QP per block for encoding. Currently
+     * VAEncDeltaQPBufferType is defined for HEVC encoding, see
+     * #VAEncDeltaQPBuffer for details
+     */
+    VAEncDeltaQPBufferType                   = 61,
```
Comment (fhvwy) [inline va/va.h:2054]: For the encoder is this meant to be an input or an output or both?
As in input it sounds like QP deltas to use when encoding, like a more complex but finer-grained way of doing ROI.  As an output it sounds like the QP deltas which got applied by BRC.

File: va/va.h [hunk 3]
```diff
@@ -2906,6 +2915,17 @@ typedef struct _VAEncMiscParameterCustomRoundingControl {
         uint32_t    value;
     }   rounding_offset_setting;
 } VAEncMiscParameterCustomRoundingControl;
+typedef struct _VAEncDeltaQPBuffer {
+    /*
+     * \brief This structure holds delta QP per block, the size of block must be queried 
+     * for the specified codec, like HEVC.
+     * Buffer size shall be sufficient to fit the slice or frame to be encoded depending on
+     * if it is a slice level or frame level encoding.
```
Comment (fhvwy) [inline va/va.h:2923]: What order are the blocks in?
I assume if encode order and raster scan order agree then it's that, but they don't match in codecs with tiles (VP9, AV1 or HEVC) when there is more than one tile horizontally.

File: va/va.h (hunk 2 above)
Comment (Tianhaol) [reply va/va.h:2054]: This is an input buffer

File: va/va_enc_hevc.h [hunk 4]
```diff
@@ -225,8 +228,14 @@ typedef union VAConfigAttribValEncHEVCBlockSizes {
...
 
+        /** Supported size of qp or delta qp block.
+         *
+         *  AP must fill in qp or delta qp buffer in unit of block.
+         */
+        uint32_t log2_qp_coding_block_size_minus3          : 2;
+
         /** Reserved for future use. */
-        uint32_t reserved                                  : 10;
+        uint32_t reserved                                  : 8;
```
[the remark this answers is not in the document]
Comment (Tianhaol) [reply va/va_enc_hevc.h:238]: It should be a better solution.

File: va/va.h (hunk 3 above)
Comment (Tianhaol) [reply va/va.h:2923]: The order is raster scan order. The buffer is filled in with delta qp  based on frame.

**Comment by Tianhaol** [comment]
@XinfengZhang @fhvwy  Could you please review this PR again?

File: va/va.h [hunk 5]
```diff
@@ -1021,6 +1021,13 @@ typedef enum {
      * VAConfigAttribValEncAV1Ext2 union.
      */
     VAConfigAttribEncAV1Ext2            = 54,
+    /** \brief delta QP per block attribute for Encoding.  Read-only.
+     *
+     * This attribute describes whether to support delta qp per block
+     * and the supported size of delta qp block. The value
+     * returned uses the VAConfigAttribValEncDeltaQpPerBlock type.
+     */
+    VAConfigAttribEncDeltaQpPerBlock    = 55,
```
Comment (XinfengZhang) [inline va/va.h:1030]: suggest to change to , perMBControl for future usage.

File: va/va.h [hunk 6]
```diff
@@ -1401,6 +1408,19 @@ typedef union _VAConfigAttribValContextPriority {
...
+    struct {
+        /** \brief whether to support dela qp per block */
+        uint32_t delta_qp_support         : 1;
+        /** \brief supported size of delta qp block */
+        uint32_t log2_delta_qp_block_size : 4;
+        /** \brief reserved bit for future, must be zero */
+        uint32_t reserved                 : 27;
+        } bits;
+        uint32_t value;
+} VAConfigAttribValEncDeltaQpPerBlock;
```
Comment (XinfengZhang) [inline va/va.h:1422]: same as above

File: va/va.h [hunk 7]
```diff
@@ -2931,6 +2960,15 @@ typedef struct _VAEncMiscParameterCustomRoundingControl {
...
 } VAEncMiscParameterCustomRoundingControl;
+typedef struct _VAEncDeltaQpPerBlockBuffer {
+    /*
+     * \brief This structure holds delta QP for each block in raster scan order.
+     * The size of block must be queried for the specified codec, like HEVC.
+     * Buffer size shall be sufficient to fit frame to be encoded.
+     * If qp buffer and delta qp buffer are provided at the same time, qp buffer will be ignored.
+     */
+    int16_t delta_qp;
+} VAEncDeltaQpPerBlockBuffer;
```
Comment (XinfengZhang) [inline va/va.h:2971]: I wonder whether it is needed. seems the description in VAEncDeltaQpPerBlockBufferType is enough, and also add more comments to describe the layout and the order, such as raw raster order? , and suppose it is a perframe block, will you add per tile or per slice control?

File: va/va.h (hunk 5 above)
Comment (Tianhaol) [reply va/va.h:1030]: Rename it to VAConfigAttribPerMBControl?

File: va/va.h (hunk 6 above)
Comment (Tianhaol) [reply va/va.h:1422]: Rename it to VAConfigAttribValEncPerMBControl

File: va/va.h (hunk 7 above)
Comment (Tianhaol) [reply va/va.h:2971]: Remove the redundant comments.  Yes, it is a per frame block. I do not add per tile or per slice control.

**Comment by Tianhaol** [comment]
@XinfengZhang Could you help review this PR?

File: va/va.h (hunk 7 above)
Comment (XinfengZhang) [reply va/va.h:2969]: how about remove the whole structure. I dont expect anybody to use the structure to allocate, access the delta qp , right?

**Review by XinfengZhang — APPROVED** [review]
I have no other concern, @fhvwy do you have other concerns

**Comment by Tianhaol** [comment]
@XinfengZhang  If remove this whole picture, how to tell users the type of delta qp? And the type also specifies the range of delta qp.

**Comment by XinfengZhang** [comment]
> @XinfengZhang If remove this whole picture, how to tell users the type of delta qp? And the type also specifies the range of delta qp.

the data length and layout could be described inside the comments of VAEncDeltaQpPerBlockBufferType

**Comment by Tianhaol** [comment]
@XinfengZhang  If there are none of concerns from hfvwy,  could you help merge this PR?

File: va/va.h [hunk 8]
```diff
@@ -1021,6 +1021,13 @@ typedef enum {
      * VAConfigAttribValEncAV1Ext2 union.
      */
     VAConfigAttribEncAV1Ext2            = 54,
+    /** \brief Settings per MB attribute for Encoding.  Read-only.
+     *
+     * This attribute describes whether to support delta qp per block
+     * and the supported size of delta qp block at present. The value
+     * returned uses the VAConfigAttribValEncDeltaQpPerBlock type.
+     */
+    VAConfigAttribEncPerMBControl       = 55,
```
Comment (XinfengZhang) [inline va/va.h:1030]: please add it into va_str.c

File: va/va.h [hunk 9]
```diff
@@ -2069,6 +2089,15 @@ typedef enum {
     */
     VAEncryptionParameterBufferType = 60,
 
+    /**
+     * \brief Encoding delta QP per block buffer
+     *
+     * This input buffer contains delta QP per block for encoding.
+     * Currently VAEncDeltaQpPerBlockBufferType is defined for HEVC encoding,
```
Comment (XinfengZhang) [inline va/va.h:2096]: better to add "the buffer only could be created and accepted when /c VAConfigAttribValEncPerMBControl delta_qp_support == 1".
please dont specify it in the common definitions, if you want to specify the bit length for each value, maybe could use some bits in VAConfigAttribValEncPerMBControl?

File: va/va.h [hunk 10]
```diff
@@ -2931,6 +2960,13 @@ typedef struct _VAEncMiscParameterCustomRoundingControl {
         uint32_t    value;
     }   rounding_offset_setting;
 } VAEncMiscParameterCustomRoundingControl;
+typedef struct _VAEncDeltaQpPerBlockBuffer {
```
Comment (XinfengZhang) [inline va/va.h:2963]: same as above

**Review by XinfengZhang — CHANGES_REQUESTED** [review]

File: va/va.h (hunk 8 above)
Comment (Tianhaol) [reply va/va.h:1030]: Yes, I add it into va_str.c

File: va/va.h (hunk 7 above)
Comment (Tianhaol) [reply va/va.h:2969]: Yes, remove this structure. But how to indicate the size and type of delta qp value.

File: va/va.h (hunk 10 above)
Comment (Tianhaol) [reply va/va.h:2963]: Remove this structure.

File: va/va.h (hunk 9 above)
Comment (Tianhaol) [reply va/va.h:2096]: Add the comments.

**Comment by Tianhaol** [comment]
@XinfengZhang Could you help review this PR?

File: va/va.h (hunk 9 above)
Comment (XinfengZhang) [reply va/va.h:2098]: Currently VAEncDeltaQpPerBlockBufferType is defined for HEVC encoding,see #VAEncDeltaQpPerBlockBuffer for details. ?? it was removed , right? ,maybe you should add detlaQP is a 16bit signed value for each block?

Comment (Tianhaol) [reply va/va.h:2098]: An int8_t is enough for hevc.  Do we define it as an int16_t for other codec?

File: va/va.h [hunk 11]
```diff
@@ -1401,6 +1408,21 @@ typedef union _VAConfigAttribValContextPriority {
...
 
+/** brief Attribute value VAConfigAttribEncPerMBControl */
+typedef union VAConfigAttribValEncPerMBControl {
+    struct {
+        /** \brief whether to support dela qp per block */
+        uint32_t delta_qp_support         : 1;
+        /** \brief supported size of delta qp block */
+        uint32_t log2_delta_qp_block_size : 4;
+        /** \brief size of delta qp in bytes*/
+        uint32_t delta_qp_size            : 3;
```
Comment (XinfengZhang) [inline va/va.h:1419]: is there better name?  it is confusing , whole buffer size of per qp value size

**Review by XinfengZhang — COMMENTED** [review]
thanks , I have no much concern, but still hope there are a better name for delta_qp_size.

File: va/va.h (hunk 11 above)
Comment (Tianhaol) [reply va/va.h:1419]: Ok,  I'll get a better name for this field.

**Review by XinfengZhang — APPROVED** [review]

**Comment by Tianhaol** [comment]
@XinfengZhang  Could you help merge it, If there are none of concerns?

File: va/va.h [hunk 12]
```diff
@@ -1401,6 +1408,21 @@ typedef union _VAConfigAttribValContextPriority {
...
 
+/** brief Attribute value VAConfigAttribEncPerMBControl */
+typedef union VAConfigAttribValEncPerMBControl {
+    struct {
+        /** \brief whether to support dela qp per block */
+        uint32_t delta_qp_support         : 1;
+        /** \brief supported size of delta qp block */
+        uint32_t log2_delta_qp_block_size : 4;
+        /** \brief size of delta qp per block in bytes*/
+        uint32_t delta_qp_per_block_size  : 3;
```
Comment (xhaihao) [inline va/va.h:1419]: How about renaming this member to delta_qp_size_in_bytes ? If so, a VA user needn't go through the comment in va.h and know what this member is used for

File: va/va.h [hunk 13]
```diff
@@ -1401,6 +1408,21 @@ typedef union _VAConfigAttribValContextPriority {
...
+        /** \brief reserved bit for future, must be zero */
+        uint32_t reserved                 : 24;
+        } bits;
```
Comment (xhaihao) [inline va/va.h:1422]: please fix the indentation.

File: va/va.h [hunk 14]
```diff
@@ -1401,6 +1408,21 @@ typedef union _VAConfigAttribValContextPriority {
...
+        uint32_t value;
+} VAConfigAttribValEncPerMBControl;
```
Comment (xhaihao) [inline va/va.h:1424]: It is 'PerMB' in the structure name, however it is 'per block' in the comment for each member. Is this structure added for ' mb' or 'block' ?

File: va/va.h (hunk 12 above)
Comment (Tianhaol) [reply va/va.h:1419]: It is a better name.  I'll rename to it.

File: va/va.h (hunk 13 above)
Comment (Tianhaol) [reply va/va.h:1422]: Ok, fix it.

File: va/va.h (hunk 14 above)
Comment (Tianhaol) [reply va/va.h:1424]: Change it to block

**Review by xhaihao — APPROVED** [review]

**Review by XinfengZhang — APPROVED** [review]

**Review by XinfengZhang — APPROVED** [review]

**Pull request merged**